src/combinatorics/quiver.lean

```diff
@@ -135,33 +144,34 @@ end path
 
 /-- A quiver is an arborescence when there is a unique path from the default vertex
     to every other vertex. -/
-class arborescence {V} (T : quiver.{v u} V) : Sort (max (u+1) v) :=
+class arborescence (V : Type u) [quiver.{v} V] : Type (max u v) :=
 (root : V)
-(unique_path : Π (b : V), unique (T.path root b))
+(unique_path : Π (b : V), unique (path root b))
 
 /-- The root of an arborescence. -/
-protected def root {V} (T : quiver V) [arborescence T] : V :=
-arborescence.root T
+def root (V : Type u) [quiver.{v} V] [arborescence V] : V :=
+arborescence.root
 
-instance {V} (T : quiver V) [arborescence T] (b : V) : unique (T.path T.root b) :=
+instance {V : Type u} [quiver.{v} V] [arborescence V] (b : V) : unique (path (root V) b) :=
 arborescence.unique_path b
 
 /-- An `L`-labelling of a quiver assigns to every arrow an element of `L`. -/
-def labelling {V} (G : quiver V) (L) := Π a b, G.arrow a b → L
+def labelling (V : Type u) [quiver.{v} V] (L : Sort*) := Π a b : V, (a ⟶ b) → L
 
-instance {V} (G : quiver V) (L) [inhabited L] : inhabited (G.labelling L) :=
+instance {V : Type u} [quiver.{v} V] (L) [inhabited L] : inhabited (labelling V L) :=
 ⟨λ a b e, default L⟩
 
-/-- To show that `T : quiver V` is an arborescence with root `r : V`, it suffices to
+/-- To show that `[quiver V]` is an arborescence with root `r : V`, it suffices to
   - provide a height function `V → ℕ` such that every arrow goes from a
     lower vertex to a higher vertex,
   - show that every vertex has at most one arrow to it, and
   - show that every vertex other than `r` has an arrow to it. -/
-noncomputable def arborescence_mk {V} (T : quiver V) (r : V)
+noncomputable def arborescence_mk {V : Type u} [quiver.{v} V] (r : V)
   (height : V → ℕ)
-  (height_lt : ∀ ⦃a b⦄, T.arrow a b → height a < height b)
-  (unique_arrow : ∀ ⦃a b c⦄ (e : T.arrow a c) (f : T.arrow b c), a = b ∧ e == f)
-  (root_or_arrow : ∀ b, b = r ∨ ∃ a, nonempty (T.arrow a b)) : arborescence T :=
+  (height_lt : ∀ ⦃a b⦄, (a ⟶ b) → height a < height b)
+  -- TODO perhaps better to use eq_to_hom here!
```

Comment:
`eq_to_hom` requires identity arrows. What we could do is define a type of all incoming arrows and say it has a unique element.

Reply:
Yes, this comment was misguided. I'll remove it. I don't think there's any harm using `heq` here.

Reply:
```suggestion
```